File: 367.java
class Solution {
    public boolean isPerfectSquare(int num) {
        // time: O(logn), space: O(1)
        int l = 1;
        int h = num;
        while (l <= h) {
            int m = l + (h - l) / 2;
            long sq = (long) m * m;
            if (sq == num) {
                return true;
            } else if (sq < num) {
                l = m + 1;
            } else {
                h = m - 1;
            }
        }
        return false;
    }
}

File: 367.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        # time: O(n), space: O(n)
        squares = []
        x = 0
        while (x ** 2) < (2 ** 31 - 1):
            squares.append(x ** 2)
            x += 1
        return num in squares